### PR TITLE
Fixed #28897 -- Fixed QuerySet.update() on querysets ordered by annotations.

### PR DESCRIPTION
**Pull request by David-Wobrock**
Fixes https://code.djangoproject.com/ticket/28897

### CODE REVIEW

File: django/db/models/query.py [hunk 1]
```diff
@@ -1159,6 +1159,9 @@ def update(self, **kwargs):
         self._for_write = True
         query = self.query.chain(sql.UpdateQuery)
         query.add_update_values(kwargs)
+        query.order_by = tuple(
+            col for col in query.order_by if col not in query.annotations
+        )
```
Comment (charettes) [inline django/db/models/query.py:1164]: One thing that makes me uneasy here is that this code would make the following preserve ordering?

```
Bar.objects.order_by(Count("m2m_foo"))
```

Should we inline ordering references in ordering instead of silently eliminating them? It feels like we should as `annotate(foo=bar).order_by('foo')` should be equivalent to `order_by(bar)`?

Comment (David-Wobrock) [reply django/db/models/query.py:1164]: Alternatively, would it make sense/be viable to entirely drop the `order_by` as suggested Mariusz in https://code.djangoproject.com/ticket/28897#comment:6 with `query.clear_ordering()`?
The behaviour would at least be deterministic, without corner cases (and it would require some additional documentation maybe).

However, I'm not sure if we need to keep the order by for an `update` call, or if we can always prune it 🤔

Comment (David-Wobrock) [reply django/db/models/query.py:1164]: Hmm, actually `Bar.objects.order_by(Count("m2m_foo"))` is already raising an exception:
```
ERROR: test_update_ordered_by_m2m_annotation (update.tests.AdvancedTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/david/django/tests/update/tests.py", line 237, in test_update_ordered_by_m2m_annotation
    Bar.objects.order_by(Count("m2m_foo")).update(x=3)
  File "/home/david/django/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/david/django/django/db/models/query.py", line 1623, in order_by
    obj.query.add_ordering(*field_names)
  File "/home/david/django/django/db/models/sql/query.py", line 2206, in add_ordering
    raise FieldError(
django.core.exceptions.FieldError: Using an aggregate in order_by() without also including it in annotate() is not allowed: Count(F(m2m_foo))
```
However, that would only be true for aggregations.

You are right for
```
Bar.objects.order_by(Abs("m2m_foo"))
```
and
```
Bar.objects.annotate(abs_id=Abs("m2m_foo")).order_by("abs_id")
```
which should have the same behaviour but have not with the current code 🤔 

-------------------

Raising an error in the appropriate case would probably make sense. Here the different cases (when we call `.update` on them):
- inline aggregation `Bar.objects.order_by(Count("m2m_foo"))` => existing raise _Using an aggregate in order_by() without also including it in annotate() is not allowed_
- annotate aggregation `Bar.objects.annotate(m2m_count=Count("m2m_foo")).order_by("m2m_count")` => **new error**, something like _Cannot update when ordering on an annotation. Try using inline order_by(Count(F(m2m_foo)))_
- inline function `Bar.objects.order_by(Abs("m2m_foo"))` => should work
- annotate function `Bar.objects.annotate(abs_id=Abs("m2m_foo")).order_by("abs_id")` => **new error**, something like _Cannot update when ordering on an annotation. Try using inline order_by(Abs(F(m2m_foo)))_

However, the user experience is probably not great for two reasons:
1. when using an annotate aggregation, the message suggests that it might work when changing the caller - which is not true (and we know it)
2. when using an annotate function, we actually could automatically translate it.

I'll try to push changes for the "more advanced" version

File: tests/update/tests.py [hunk 2]
```diff
@@ -225,6 +225,33 @@ def test_update_with_joined_field_annotation(self):
...
+            "UPDATE...ORDER BY syntax for MySQL/MariaDB does not support ordering"
+            "related fields"
+        ),
+    )
+    def test_update_ordered_by_m2m_annotation(self):
+        foo = Foo.objects.create(target="bbb")
+        Bar.objects.create(foo=foo)
+
+        Bar.objects.order_by(Abs("m2m_foo")).update(x=2)
+        self.assertEqual(Bar.objects.get().x, 2)
```
Comment (felixxm) [inline tests/update/tests.py:250]: This works without the patch so we should move it to a separate commit and test method.

File: django/db/models/query.py [hunk 3]
```diff
@@ -1159,6 +1159,23 @@ def update(self, **kwargs):
         self._for_write = True
         query = self.query.chain(sql.UpdateQuery)
         query.add_update_values(kwargs)
+
+        # Try inlining annotations for order_by.
```
Comment (felixxm) [inline django/db/models/query.py:1163]:
```suggestion
        # Inline annotations in order_by(), if possible.
```

File: django/db/models/query.py [hunk 4]
```diff
@@ -1159,6 +1159,23 @@ def update(self, **kwargs):
...
+        new_order_by = []
+        for col in query.order_by:
+            annotation = query.annotations.get(col)
+            if annotation:
+                if getattr(annotation, "contains_aggregate", False):
+                    raise exceptions.FieldError(
+                        f"Cannot update when ordering on an aggregation annotation: "
```
Comment (felixxm) [inline django/db/models/query.py:1170]: `by` :thinking: 
```suggestion
                        f"Cannot update when ordering by an aggregate: "
```

File: django/db/models/query.py [hunk 5]
```diff
@@ -1159,6 +1159,23 @@ def update(self, **kwargs):
...
+                        f"{annotation}"
+                    )
+                else:
+                    new_order_by.append(annotation)
```
Comment (felixxm) [inline django/db/models/query.py:1174]:
```suggestion
                new_order_by.append(annotation)
```

File: tests/update/tests.py [hunk 6]
```diff
@@ -225,6 +225,33 @@ def test_update_with_joined_field_annotation(self):
...
+        with self.assertRaisesMessage(FieldError, msg):
+            Bar.objects.annotate(m2m_count=Count("m2m_foo")).order_by(
+                "m2m_count"
+            ).update(x=2)
+
+    @unittest.skipIf(
+        connection.vendor == "mysql",
+        (
+            "UPDATE...ORDER BY syntax for MySQL/MariaDB does not support ordering"
+            "related fields"
```
Comment (felixxm) [inline tests/update/tests.py:242]: Is this something that we can support? so is it a backend or database limitation?

Database limitation, I suppose.

**Review by felixxm — COMMENTED** [review]
@David-Wobrock Thanks :+1: I left small comments.

@charettes Do you want to take another look?

File: django/db/models/query.py [hunk 7]
```diff
@@ -1169,6 +1169,20 @@ def update(self, **kwargs):
         self._for_write = True
         query = self.query.chain(sql.UpdateQuery)
         query.add_update_values(kwargs)
+
+        # Inline inlining annotations in order_by, if possible.
```
Comment (charettes) [inline django/db/models/query.py:1173]:
```suggestion
        # Inline annotations in order_by, if possible.
```

**Review by charettes — COMMENTED** [review]
Changes look great to me, thank you for the tweaks @David-Wobrock!

**Comment by David-Wobrock** [comment]
Thanks for the reviews 🙏

**Comment by felixxm** [comment]
@David-Wobrock Thanks :+1:
